library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleQueue.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.exoplayer2.source.hls;
+
+import androidx.annotation.Nullable;
+import com.google.android.exoplayer2.C;
+import com.google.android.exoplayer2.Format;
+import com.google.android.exoplayer2.drm.DrmInitData;
+import com.google.android.exoplayer2.drm.DrmSessionManager;
+import com.google.android.exoplayer2.metadata.Metadata;
+import com.google.android.exoplayer2.metadata.id3.PrivFrame;
+import com.google.android.exoplayer2.source.SampleQueue;
+import com.google.android.exoplayer2.source.UnexpectedDiscontinuityException;
+import com.google.android.exoplayer2.upstream.Allocator;
+import com.google.android.exoplayer2.util.Assertions;
+import com.google.android.exoplayer2.util.Log;
+import com.google.android.exoplayer2.util.MediaSourceEventDispatcher;
+import com.google.android.exoplayer2.util.TimestampAdjuster;
+import java.util.Map;
+
+/**
+ * Extend base SampleQueue to add HLS specific processing of the samples, including:
+ *
+ * <ul>
+ *   <li>segment time boundary checks on timestamps of committed samples</li>
+ *   <li>cleaning the {@link Format#metadata} to avoid excessive format changes</li>
+ * </ul>
+ *
+ * The timestamp check verifies that the adjusted sample time (via {@link TimestampAdjuster}) does not
+ * fall outside of a set percentage ({@link #MAX_TIMESTAMP_DEVIATION_PERCENTAGE}) of the time
+ * boundaries of the segment as expressed by the segment duration ((@link HlsMediaChunk#endTimeUs} -
+ * {@link HlsMediaChunk#startTimeUs}).  This is loosely mandated by the Pantos spec and checked by
+ * Apple's mediastreamvalidator.
+ *
+ */
+public class HlsSampleQueue extends SampleQueue {
```

Comment:
Can this stay where `FormatAdjustingSampleQueue` was (i.e., as an inner class)? We definitely don't want it to be part of the ExoPlayer library's actual API surface, and I don't think enough is really being added to justify splitting it into its own file.

Reply:
IMHO, breaking up big classes is a good thing (less conflicted merges, etc), I can make it package private and exclude it from the JavaDoc, but I do think it is a good idea to make it a standalone class.  Pretty much equivalent from the VM's view to a static inner class.

If I can exclude it from the package classes are you ok with this?  Otherwise, sure I'll move it to an inner class.

Reply:
It's not a great idea to split it out whilst simultaneously modifying it, because it makes the change harder to review (or look at later to see what changed). If it's to be split out, that should IMO be a separate change after this one, that's a pure no-op move rather than one that also changes behavior.

Reply:
Ah. Yes, I agree 100%, I"ll move it back and reduce the diffs.

Reply:
@ojw28 In this commit ccf2ba3e1 I moved the class back as an inner class and tried to remove the diffs.   Once we finish the pull up and agree it is merge ready I will rebase out the intermediate states to further reduce the changes to `HlsSampleStreamWrapper` so the y are clear.

Hope this helps.